File: app/src/main/java/com/agrogames/islandsofwar/render/impl/UnitList.java
package com.agrogames.islandsofwar.render.impl;

import android.os.Build;

import com.agrogames.islandsofwar.engine.abs.common.Cell;
import com.agrogames.islandsofwar.engine.abs.transport.TransportUnit;
import com.agrogames.islandsofwar.engine.abs.weapon.IWeapon;
import com.agrogames.islandsofwar.ui.abs.Element;
import com.agrogames.islandsofwar.ui.abs.ElementType;
import com.agrogames.islandsofwar.ui.abs.UI;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class UnitList {
    private final UI ui;
    private final float x;
    private final List<Element> elements = new ArrayList<>();
    private TransportUnit currentUnit;

    public UnitList(UI ui, float x) {
        this.ui = ui;
        this.x = x;
    }

    public void setUnits(TransportUnit[] units){
        if(currentUnit != null && !Arrays.asList(units).contains(currentUnit)) currentUnit = null;
        for (Element e : elements){
            ui.deleteElement(e);
        }
        List<TransportUnit> unitTypes = new ArrayList<>();
        for(TransportUnit unit : units){
            if(!unitTypes.contains(unit)){
                unitTypes.add(unit);
            }
        }
        float y = 7.5f;
        for (TransportUnit unit : unitTypes){
            Element button = ui.createElement(ElementType.Button, x, y, 1.2f, 1.2f, "button_background");
            button.onClick(() -> {
                if(currentUnit == null || currentUnit != unit) {
                    clearUnits();
                    currentUnit = unit;
                    setUnits(units);
                } else {
                    clearUnits();
                    currentUnit = null;
                    setUnits(units);
                }
                return null;
            });
            elements.add(button);

            Element e = ui.createElement(ElementType.Button, x, y, -1, -1, TextureMapper.join(unit.example.getTexture(),
                    unit == currentUnit ? "selected" : "normal"));
            e.setRenderInBorders(false);
            elements.add(e);

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                for (IWeapon weapon : unit.create.apply(new Cell(0, 0)).getWeapons()){
                    Element e1 = ui.createElement(ElementType.Button, x + weapon.getRelativeLocation().x, y + weapon.getRelativeLocation().y, -1, -1,
                            TextureMapper.join(weapon.getTexture()));
                    e1.setRenderInBorders(false);
                    elements.add(e1);
                }
            }
            y -= 1;
        }
    }

    public TransportUnit getCurrentUnit(){
        return currentUnit;
    }

    public void clearUnits() {
        clearUnits(true);
    }

    public void clearUnits(boolean replaceCurrent){
        if(replaceCurrent)
            currentUnit = null;
        for (Element e : elements){
            ui.deleteElement(e);
        }
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/render/impl/UnitList.kt
package com.agrogames.islandsofwar.render.impl

import com.agrogames.islandsofwar.engine.abs.transport.TransportUnit
import com.agrogames.islandsofwar.ui.impl.Element
import com.agrogames.islandsofwar.ui.impl.ElementType
import com.agrogames.islandsofwar.ui.impl.UI

class UnitList(private val ui: UI, private val x: Float) {
    private val elements: MutableList<Element> = ArrayList()
    var currentUnit: TransportUnit? = null
        private set

    fun setUnits(units: Array<TransportUnit>) {
        if (currentUnit != null && !units.contains(currentUnit)) currentUnit = null
        for (e in elements) {
            ui.deleteElement(e)
        }
        var y = 7.5f
        for (unit in setOf(*units)) {
            elements.add(ui.createElement(Element(ElementType.Button, x, y, 1.2f, 1.2f, "ui/button_background", {
                if (currentUnit == null || currentUnit !== unit) {
                    clearUnits()
                    currentUnit = unit
                    setUnits(units)
                } else {
                    clearUnits()
                    currentUnit = null
                    setUnits(units)
                }
            })))
            elements.add(ui.createElement(Element(
                ElementType.Button, x, y, 1.2f, 1.2f, TextureMapper.join(unit.example.texture,
                    if (unit == currentUnit) "selected" else "normal"), renderInBorders = false)))
            for (weapon in unit.example.weapons) {
                val e1 = ui.createElement(Element(
                    ElementType.Button, x + weapon.relativeLocation.x, y + weapon.relativeLocation.y, -1f, -1f,
                        TextureMapper.join(weapon.texture)))
                e1.renderInBorders = false
                elements.add(e1)
            }
            y -= 1f
        }
    }

    fun clearUnits(replaceCurrent: Boolean = true) {
        if (replaceCurrent) currentUnit = null
        for (e in elements) {
            ui.deleteElement(e)
        }
    }
}